ext4/dmverity/dmverity.go

```diff
@@ -0,0 +1,175 @@
+package dmverity
+
+import (
+	"bytes"
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/pkg/errors"
+
+	"github.com/Microsoft/hcsshim/ext4/internal/compactext4"
+)
+
+const (
+	blockSize = compactext4.BlockSize
+)
+
+var (
+	salt                      = bytes.Repeat([]byte{0}, 32)
+	ErrSuperBlockReadFailure  = errors.New("failed to read dm-verity super block")
+	ErrSuperBlockParseFailure = errors.New("failed to parse dm-verity super block")
+	ErrRootHashReadFailure    = errors.New("failed to read dm-verity root hash")
+)
+
+type dmveritySuperblock struct {
+	/* (0) "verity\0\0" */
+	Signature [8]byte
+	/* (8) superblock version, 1 */
+	Version uint32
+	/* (12) 0 - Chrome OS, 1 - normal */
+	HashType uint32
+	/* (16) UUID of hash device */
+	UUID [16]byte
+	/* (32) Name of the hash algorithm (e.g., sha256) */
+	Algorithm [32]byte
+	/* (64) The data block size in bytes */
+	DataBlockSize uint32
+	/* (68) The hash block size in bytes */
+	HashBlockSize uint32
+	/* (72) The number of data blocks */
+	DataBlocks uint64
+	/* (80) Size of the salt */
+	SaltSize uint16
+	/* (82) Padding */
+	_ [6]byte
+	/* (88) The salt */
+	Salt [256]byte
+	/* (344) Padding */
+	_ [168]byte
+}
+
+// VerityInfo is minimal exported version of dmveritySuperblock
+type VerityInfo struct {
+	// Offset in blocks on hash device
+	HashOffsetInBlocks int64
+	// Set to true, when dm-verity super block is also written on the hash device
+	SuperBlock    bool
+	RootDigest    string
+	Salt          string
+	Algorithm     string
+	DataBlockSize uint32
+	HashBlockSize uint32
+	DataBlocks    uint64
+}
+
+func Tree(data []byte) []byte {
```

Comment:
not in scope of this PR, but this should probably be some sort of a `Reader` and we'd read the VHD data sequentially in blocks, rather than everything into memory?

Reply:
Perhaps. I went with the "easiest thing to get working". I'm not really sure whether it should be or not.

Reply:
If its a Reader then you need to write the tree somewhere... for the most common use case that would mean, writing the tree somewhere to save a probably small amount of memory so you can read the tree from that location to write it to its final location. Not really sure that is a win. It's more complicated, unlike to have much memory pressure and would be doing extra IO.